Add resource chapter documenting buffers and bindings (#344)
microsoft/hlsl-specs: specs/language/resources.tex

but must be resolvable to a unique global buffer.

\begin{HLSL}
  Buffer<int> grobuf;
  RWBuffer<int> grwbuf;
  void main() {
    Buffer lrobuf<int> = grobuf;
    RWBuffer lrwbuf<int> = grwbuf;
  }
\end{HLSL}
Buffer operands to the assignment operator must have the same element type.

\Sub{Dimensions}{Resources.tybufs.dims}

The size of a typed buffer can be retrieved using the \texttt{GetDimensions} method.
\begin{HLSL}
void GetDimensions(out uint width) const;
\end{HLSL}

This returns the full legnth, in elements of the buffer through the \texttt{width} out parameter.

\Sub{Element Access}{Resources.tybufs.access}

The contents of typed buffers can be retrieved using the \texttt{Load} methods
or the subscript operator.

\begin{HLSL}
 T Load(in int pos) const;
 T operator[](uint pos) const;
\end{HLSL}

These each return the element at the given position \texttt{pos} of the type specified in the buffer definition.

An additional \texttt{Load} method returns the indicated element just as the other,
but also takes a second \texttt{status} parameter that returns information about the accessed resource.
\begin{HLSL}
 T Load(in int pos, out uint status) const;
\end{HLSL}

> The \texttt{status} parameter returns an indication of whether the value retrieved accessed a mapped tile
> resource. This parameter is interpretted using the \texttt{CheckAccessFullyMapped}
> built-in function. If tiled resources are enabled, it returns true if the value was accessed from a mapped
> tile resource and false otherwise. If tiled resourced are not enabled, it will return true.

Reviewer: Here's my attempt at this:

```suggestion
The \texttt{status} parameter returns an indication of whether all of the retrieved value came from fully mapped parts of the resource.  This parameter must be passed to the built-in function \texttt{CheckAccessFullyMapped} in order to interpret it.  This function returns true is the value was accessed from a fully committed resource, or from fully mapped pages of a sparse resource. If any part of the return value was from unmapped memory then false is returned.
```

Author: I took much of this text, but ended up splitting out the `CheckAccessFullyMapped` built-in to its own section where I could go into slightly more depth with less disruption to the explanation of Loads.

bit repesantations in addition to ignoring floating point rules surrounding NaN
and infinite values.

\Sec{Structured Buffers}{Resources.stbufs}

Structured buffers are raw buffers with associated types that facilitate
indexing.
Unlike typed buffers, they perform no format conversions.
The buffer contents is treated as raw data and the casting to the given types
is done bit for bit.
Structured buffers can be defined with scalar, vector, matrix, or user-defined
struct elements.

Structured buffers can be read-only, writable, appendable, or consumable.
Writable buffers can have their elements assigned in arbitrary locations.
Append structured buffers can only have elements added to the end.
Consume structured buffers can only have elements removed from the end.

\begin{HLSL}
template <typename T>
 class StructuredBuffer {
 public:
   StructuredBuffer();
   StructuredBuffer(StructuredBuffer buf);
   StructuredBuffer operator=(StructuredBuffer buf);

   void GetDimensions(out uint count, out uint stride) const;

   // Element access.
   T operator[](int pos) const;
   T Load(in int pos) const;
   T Load(in int pos, out uint status) const;
};

>  class RWStructuredBuffer {

Reviewer: Should this derive from StructuredBuffer?  And be templated?

 public:
   RWStructuredBuffer();
   RWStructuredBuffer(RWStructuredBuffer buf);
   RWStructuredBuffer operator=(RWStructuredBuffer buf);

   // Element assignment.
   T &operator[](int pos);

   // Hidden counter increment/decrement.
   uint IncrementCounter();
   uint DecrementCounter();
};

template <typename T>
 class AppendStructuredBuffer {

>    AppendStructuredBuffer();

Reviewer: missing public

   AppendStructuredBuffer(AppendStructuredBuffer buf);
   AppendStructuredBuffer operator=(AppendStructuredBuffer buf);

   void GetDimensions(out uint count, out uint stride) const;

   void Append(in T value);
};

template <typename T>
 class ConsumeStructuredBuffer {
   ConsumeStructuredBuffer();
   ConsumeStructuredBuffer(ConsumeStructuredBuffer buf);
   ConsumeStructuredBuffer operator=(ConsumeStructuredBuffer buf);

   void GetDimensions(out uint count, out uint stride) const;

>    void Consume();

Reviewer: ```suggestion
   T Consume();
```

  \define{bind-number}\br
  digit\br
  bind-number digit\br
\end{grammar}

The register type indicates whether the binding is for a read-only (\texttt{t}), a writable (\texttt{u}),
a constant buffer (\texttt{b}), or a sampler (\texttt{s}).
The register bind number indicates the register number at which the resource variable begins.
The optional second parameter indicates the virtual register space that the register belongs to.

Sample syntax of virtual register binding attributes:
\begin{HLSL}
  Buffer<float> robuf : register(t0, space0);
  RWBuffer<float> rwbuf : register(u0, space0);
  SamplerState samp : register(s0, space0);
  cbuffer cbuf : register(b0, space0) { float f; };
\end{HLSL}

Each resource must have a unique register. Different resources cannot occupy the same register,
but read-only and writable buffers have separate namespaces and different logical register spaces have independent
sets of register numbers. Meaning the following is all allowed:
\begin{HLSL}
  Buffer<float> mybuf : register(t0, space0);
  Buffer<float> yourbuf : register(t0, space1);
  RWBuffer<float> hisbuf : register(u0, space0);
  RWBuffer<float> herbuf : register(u0, space1);
\end{HLSL}

> Resource aggregates structs containing multiple resources types or arrays of resource types or structs containing one or more resource types.

Reviewer: > Resource aggregates structs containing multiple resources types or arrays of resource types or structs containing one or more resource types.

I might be tired, but I'm having a hard time parsing this one.

Its contents are indexed by typed access to each element
Types may have their formats converted upon load.

Template types can be any type that totals or can be padded to 16 bytes.

All typed buffers can be read through subscript operators or Load methods.
Writable typed buffers can be written through subscript operators.

Where \texttt{T} can be any arithmetic type \ref{Basic.types.arithmetic}
or a vector with a maximum of four elements containing such an arithmetic type.
The total size of a single contained element must be less than 16 bytes.

Typed buffers perform format conversions on load such that the underlying data
gets converted to the destination type.

\begin{HLSL}
template <typename T = float4>
 class Buffer {
 public:
   Buffer();
   Buffer(const Buffer &buf);
   Buffer operator=(Buffer buf);

   void GetDimensions(out uint width) const;

   // Element access.
   T operator[](int index) const;
   T Load(in int index) const;
   T Load(in int index, out uint status) const;
};

template <typename T>
 class RWBuffer : public Buffer {
 public:
   RWBuffer();
   RWBuffer(RWBuffer buf);
   RWBuffer operator=(RWBuffer buf);

   // element access

>    T &operator[](int n);

Reviewer 2: Should this be `T &operator[](size_t n);`?

Along the same lines, should the type of `index` argument of `Buffer::Load(...)` methods above be `size_t` instead of `int`?

Reviewer 2: > Should this be `T &operator[](size_t n);`?
> 
> Along the same lines, should the type of `index` argument of `Buffer::Load(...)` methods above be `size_t` instead of `int`?

Never mind. It would contradict the [public API](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-object-buffer-load)

\begin{HLSL}
 T Load(in int index) const;
 T operator[](int index) const;
\end{HLSL}

These each return the element of the buffer's type at the given index \texttt{index}.

An additional \texttt{Load} method returns the indicated element just as the other,
but also takes a second \texttt{status} parameter that returns information about the accessed resource.
\begin{HLSL}
 T Load(in int index, out uint status) const;
\end{HLSL}

The \texttt{status} parameter returns an indication of whether all of the retrieved value
came from fully mapped parts of the resource.
This parameter must be passed to the built-in function \texttt{CheckAccessFullyMapped} (\ref{Resources.mapcheck})
in order to interpret it.

Writable RWStructuredBuffers have an additional subscript operator that allows assignment to a structure element of the buffer.
It behaves as if it returns a reference to that element.
\begin{HLSL}
 T &operator[](int index);
\end{HLSL}

\Sub{Counter Manipulation}{Resources.stbufs.counter}

RWStructuredBuffers may have a hidden counter that can be incremented and decremented using methods.

\begin{HLSL}
   uint IncrementCounter();
   uint DecrementCounter();
\end{HLSL}

> Increment or decrements the hidden counter associated with the RWStructuredBuffer.

Reviewer 3: What is the return value?

  \define{bind-number}\br
  digit\br
  bind-number digit\br
\end{grammar}

The register type indicates whether the binding is for a read-only (\texttt{t}), a writable (\texttt{u}),
a constant buffer (\texttt{b}), or a sampler (\texttt{s}).
The register bind number indicates the register number at which the resource variable begins.
The optional second parameter indicates the virtual register space that the register belongs to.

Sample syntax of virtual register binding attributes:
\begin{HLSL}
  Buffer<float> robuf : register(t0, space0);
  RWBuffer<float> rwbuf : register(u0, space0);
  SamplerState samp : register(s0, space0);
  cbuffer cbuf : register(b0, space0) { float f; };
\end{HLSL}

Each resource must have a unique register. Different resources cannot occupy the same register,
but read-only and writable buffers have separate namespaces and different logical register spaces have independent
sets of register numbers. Meaning the following is all allowed:
\begin{HLSL}
  Buffer<float> mybuf : register(t0, space0);
  Buffer<float> yourbuf : register(t0, space1);
  RWBuffer<float> hisbuf : register(u0, space0);
  RWBuffer<float> herbuf : register(u0, space1);
\end{HLSL}

> A aggregate resource type can be:

Reviewer: ```suggestion
An aggregate resource type can be:
```

Author: Done

\begin{itemize}
\item a struct containing one or more resources types
\item an array of resource types
\item an array or struct containing either of the above
\end{itemize}

When aggregate resource types have register annotations,

> they occupy the first register they specify as well as however many additional sequential registers

Reviewer: ```suggestion
they occupy the first register they specify, as well as however many additional sequential registers
```

the aggregate requires to assign a register to each of its elements that correspond to the given register type.
Multiple register annotations can be placed on aggregates involving structs where each corresponds to a distinct
register type that will apply to all the registers of that type in the aggregate.

\begin{note}
Resource types contained in structs are only allocated registers when they are explicitly used.

> This includes elements of arrays of resources as such array elements must be indexed with literals.

Reviewer: ```suggestion
This includes elements of arrays of resources, as such array elements must be indexed with literals.
```

Author: I also don't think this comma is needed. "As" serves the same purpose as "since", which I will replace it with for similar reasons as above. 

This says that if it is a subordinate clause following the main close, no comma is needed. It is needed when it is a subordinate clause that precedes the main clause e.g. "Since such array elements must be indexed with literals, this includes elements of arrays of resources."

\Ch{Resource Types}{Resources}

Resources are built-in intangible types representing memory with an external interface.

These take the form of Typed Buffers, Raw Buffers, Textures, and Samplers.
Buffer and Texture types can be read-only or writable.
Read-only buffers and textures types constitute Shader Resource Views (SRVs) of the external memory.

> Writable buffers and textures types constitute Unordered Access Views (UAVs) of the external memory.

Author: Removed them.

\Sec{Typed Buffers}{Resources.tybuf}

The typed buffer class template represents a resource whose format can be converted upon load.

Template types can be any type that totals or can be padded to 16 bytes.

All typed buffers can be read through subscript operators or Load methods.
Writable typed buffers can be written through subscript operators.

> Typed Buffers are declared with the built-in templates:

Author: Should have been removed

   Buffer();
   Buffer operator=(Buffer buf);

   void GetDimensions(out uint width);

   // element access
   T operator[](uint pos);
   T Load(in int pos);
   T Load(in int pos, out uint status);
};
\end{HLSL}

\Sub{Constructors}{Resources.rotybufs.ctrs}

Buffers can be defined with explicit or implicit template parameter types.
\begin{HLSL}
  Buffer buf1;
  Buffer<> buf2;
  Buffer<float4> buf3;
\end{HLSL}
Since \texttt{float4} is the default type, all of these definitions are equivalent.

When defined at global scope, Buffers are bound to externally-defined backing stores
using the explicit binding location if provided or the implicit binding if not.

When defined at local scope, Buffers represent local references
that can be associated with global buffers when assigned,
but must be resolvable to a unique global buffer.

\begin{HLSL}
  Buffer<int> gbuf;
  void doit() {
    Buffer lbuf<int> = gbuf;
  }
\end{HLSL}
Buffer operands to the assignment operator must have the same element type.

\Sub{Dimensions}{Resources.rotybufs.dims}

The size of a Buffer can be retrieved using the \texttt{GetDimensions} method.
\begin{HLSL}
void GetDimensions(out uint size);
\end{HLSL}

> This returns the full size of the buffer in bytes through the \texttt{width} out parameter.

Author: changed

\Sub{Element Access}{Resources.rotybufs.access}

The contents of Buffers can be retrieved using the \texttt{Load} methods
or the subscript operator.

\begin{HLSL}
 T Load(in int pos);

>  T operator[](uint pos);

Author: That's what the documentation says, but it turns out that a -1 makes it through for the subscript in DXC. I don't know how that would be useful though.

>  T operator[](uint pos);

Reviewer: Yup, I was suggesting that we should change the spec here, only call out that this somewhat surprising definition is actually what is intended and isn't a mistake.

\end{HLSL}

These each return the element at the given position \texttt{pos} of the type specified in the buffer definition.

An additional \texttt{Load} method returns the indicated element just as the other,
but also takes a second \texttt{status} parameter that returns information about the accessed resource.
\begin{HLSL}
 T Load(in int pos, out uint status);
\end{HLSL}

The \texttt{status} parameter returns an indication of whether the value retrieved accessed a mapped tile
resource on tile-based platforms. This parameter is interpretted using the \texttt{CheckAccessFullyMapped}
built-in function. If tiled resources are enabled, it returns true if the value was accessed from a mapped

> tile resource and false otherwise. If tiled resourced are not enabled, it will return true.

Author: Left over from when I didn't follow this clearly. removed

\Sec{Writable Typed Buffers}{Resources.rwtybufs}

\begin{HLSL}
template <typename T>
 class RWBuffer {

   RWBuffer();
   RWBuffer operator=(RWBuffer buf);

   void GetDimensions(out uint size);

   // element access
   T operator[](uint n);
   T Load(in int pos);
   T Load(in int pos, out uint status);
};
\end{HLSL}

\Sub{Constructors}{Resources.rwtybufs.ctrs}

RWBuffers must be defined with explicit template parameter types.
\begin{HLSL}
  RWBuffer<float4> buf;
\end{HLSL}

> When defined at global scope, RWBuffers are bound to externally-defined backing stores

Reviewer 3: I think the main difference between `RWBuffer` and `Buffer` is that `RWBuffer` has a non-const `operator[]` that returns a `T&`.

That isn't captured here, but we should capture it and probably describe `RWBuffer` in terms of additions to `Buffer`. In the [Work Graph's spec](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0018-work-graphs.md), I had described the record types with base classes that didn't actually exist as a way of describing the composition of interfaces. That could be a useful tool for resources too.

> When defined at global scope, RWBuffers are bound to externally-defined backing stores

Author: I went back and forth on this. I think we've tied ourselves in knots trying to consolidate common sections of documentation in the past and I noted that the C++ spec repeats a lot of things. Merging information between read-only and writable types might be manageable though. I'll take a pass at it. 

Initially, I'm more inclined to merge the sections and just annotated cases where it applies to RW buffers only. Does that sound okay?

An additional \texttt{Load} method returns the indicated element just as the other,
but also takes a second \texttt{status} parameter that returns information about the accessed resource.
\begin{HLSL}
 T Load(in int pos, out uint status);
\end{HLSL}

The \texttt{status} parameter returns an indication of whether the value retrieved accessed a mapped tile
resource on tile-based platforms. This parameter is interpretted using the \texttt{CheckAccessFullyMapped}
built-in function. If tiled resources are enabled, it returns true if the value was accessed from a mapped
tile resource and false otherwise. If tiled resourced are not enabled, it will return true.

Writable buffers have an additional subscript operator that allows assignment to an element of the buffer.
It behaves as if it returns a reference to that element.
\begin{HLSL}
 T &operator[](uint pos);
\end{HLSL}

\Sec{Raw Buffer Types}{Resources.rawbufs}

Raw buffers are either ByteAddressBuffers or StructuredBuffers.
ByteAddressBuffers enable per-byte access to the raw data while
StructuredBuffers have an associated structure type that determines how they are

> indexed. This type can be a scalar, vector, matrix, or user-defined struct.

Author: I added something there that I hope is equivalent.

   void GetDimensions(out uint size);

   // element access
   uint Load(in uint byteOffset)
   uint2 Load2(in uint byteOffset)
   uint3 Load3(in uint byteOffset)
   uint4 Load4(in uint byteOffset)
   template<typename T>
   T Load(in uint byteOffset);

   uint Load(in uint byteOffset, out uint status)
   uint2 Load2(in uint byteOffset, out uint status)
   uint3 Load3(in uint byteOffset, out uint status)
   uint4 Load4(in uint byteOffset, out uint status)
   template<typename T>
   T Load(in uint byteOffset, out uint status);
};
\end{HLSL}

\Sub{Constructors}{Resources.robabufs.ctrs}

Since their contents are viewed as raw bytes, ByteAddressBuffers are defined as a simple variable definition.
\begin{HLSL}
  ByteAddressBuffer buf;
\end{HLSL}

When defined at global scope, ByteAccessBuffers are bound to externally-defined backing stores
using the explicit binding location if provided or the implicit binding if not.

When defined at local scope, ByteAccessBuffers represent local references
that can be associated with global ByteAccessBuffers when assigned,
but must be resolvable to a unique global buffer.

\begin{HLSL}
  ByteAddressBuffer gbuf;

>   void doit(bool b) {

Reviewer: > void doit(bool b)

Wondering why there's the boolean argument in this example?

Author: In error from when I had a more involved example. Removed.

   void InterlockedMax64(in uint pos, in uint64_t value)
   void InterlockedMax64(in uint pos, in uint64_t value, out uint64_t original)

   // 64-bit integer atomic compare/exchange operations
   void InterlockedCompareStore64(in uint pos, in uint64_t compare, in uint64_t value);
   void InterlockedExchange64(in uint pos, in int64_t value, out int64_t original);
   void InterlockedCompareExchange64(in uint pos, in uint64_t compare,
                                      in uint64_t value, out int64_t original);

   // 32-bit float atomic compare/exchange operations
   void InterlockedCompareStoreFloatBitwise(in uint byteOffest,
                                             in float compare, in float value);
   void InterlockedExchangeFloat(in uint byteOffest, in float value,
                                  out float original);
   void InterlockedCompareExchangeFloatBitwise(in uint byteOffest,
                                                in float compare,
                                                in float value,
                                                out float original);
};
\end{HLSL}

\Sub{Constructors}{Resources.rwbabufs.ctrs}

> Since their contents are viewed as raw bytes, RWByteAddressBuffers are defined as a simple variable definition.

Reviewer: Wonder if "without any type parameters" might be more specific than "simple variable definition".

Author: Replaced

using the explicit binding location if provided or the implicit binding if not.

When defined at local scope, ConsumeStructuredBuffers represent local references
that can be associated with global buffers when assigned,
but must be resolvable to a unique global buffer.

\begin{HLSL}
  ConsumeStructuredBuffer<int3x3> gbuf;
  void doit() {
    ConsumeStructuredBuffer lbuf<int3x3> = gbuf;
  }
\end{HLSL}
ConsumeStructuredBuffer operands to the assignment operator must have the same element type.

\Sub{Dimensions}{Resources.cnstbufs.dims}

The structure count and stride of a ConsumeStructuredBuffer can be retrieved using the \texttt{GetDimensions} method.
\begin{HLSL}
void GetDimensions(out uint count, out uint stride);
\end{HLSL}

This returns number of structured elements of the buffer through the \texttt{count} out parameter
and the size of each element in bytes through the \texttt{stride} out parameter.

\Sub{Consume}{Resources.cnstbufs.consume}

\begin{HLSL}
   void Consume();
\end{HLSL}

Removes and returns a value of the template parameter types from the end of the ConsumeStructuredBuffer.

\Sec{Constant Buffers}{Resources.cnbuf}

\Sec{Samplers}{Resources.samp}

> \Sec{Resource Binding}{Resources.binding}

Reviewer: This needs to have a discussion about bindings on UDT's.

Author: Added

\Ch{Resource Types}{Resources}

Resources are built-in intangible types representing memory with an external interface.

These take the form of Typed Buffers, Raw Buffers, Textures, and Samplers.
Buffer and Texture types can be read-only or writable.

\Sec{Typed Buffers}{Resources.tybuf}

The typed buffer class template represents a one-dimensional resource containing an array of a single given type.
Its contents are indexed by typed access to each element
Types may have their formats converted upon load.

Template types can be any type that totals or can be padded to 16 bytes.

All typed buffers can be read through subscript operators or Load methods.
Writable typed buffers can be written through subscript operators.

Where \texttt{T} can be any arithmetic type \ref{Basic.types.arithmetic}
or a vector with a maximum of four elements containing such an arithmetic type.
The total size of a single contained element must be less than 16 bytes.

Typed buffers perform format conversions on load such that the underlying data
gets converted to the destination type.

\Sec{Read-only Typed Buffers}{Resources.rotybufs}

\begin{HLSL}
template <typename T = float4>
 class Buffer {

>    Buffer();

Reviewer 3: We also need a copy constructor.

Reviewer: We use references elsewhere and when we do add references to the language I think it'll mean that this documentation just becomes correct.  

I think that the copy constructors should take const references though.

   Buffer operator=(Buffer buf);

   void GetDimensions(out uint width);

   // element access
   T operator[](uint pos);
   T Load(in int pos);

>    T Load(in int pos, out uint status);

Reviewer 3: Some (maybe all) of these methods should probably be `const`.

   // Element access.
   T operator[](uint pos) const;
   T Load(in int pos) const;
   T Load(in int pos, out uint status) const;
};

template <typename T>
 class RWBuffer : public Buffer {
 public:
   RWBuffer();
   RWBuffer(RWBuffer buf);
   RWBuffer operator=(RWBuffer buf);

   // element access
   T &operator[](uint n);
};
}
\end{HLSL}

\Sub{Constructors}{Resources.tybufs.ctrs}

Read-only Buffers can be defined with explicit or implicit template parameter types.
\begin{HLSL}
  Buffer buf1;
  Buffer<> buf2;
  Buffer<float4> buf3;
\end{HLSL}
Since \texttt{float4} is the default type, all of these definitions are equivalent.

RWBuffers must be defined with explicit template parameter types.
\begin{HLSL}
  RWBuffer<float4> buf;
\end{HLSL}

When defined at global scope, typed buffers are bound to externally-defined backing stores
using the explicit binding location if provided or the implicit binding if not.

When defined at local scope, typed buffers represent local references
that can be associated with global buffers when assigned,
but must be resolvable to a unique global buffer.

\begin{HLSL}
  Buffer<int> grobuf;
  RWBuffer<int> grwbuf;
  void main() {

>     Buffer lrobuf<int> = grobuf;
>     RWBuffer lrwbuf<int> = grwbuf;

Reviewer: ```suggestion
    Buffer<int> lrobuf = grobuf;
    RWBuffer<int> lrwbuf = grwbuf;
```

Author: I fixed this location and one other where I'd somehow gotten it out of order.
